File: sdk/java/examples/TestSource.java
package com.bundlebase.sdk.examples;

import com.bundlebase.sdk.*;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.BigIntVector;
import org.apache.arrow.vector.VarCharVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.Schema;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Test source function used by Rust integration tests.
 */
public class TestSource implements Connector {
    private final BufferAllocator allocator = new RootAllocator();

    @Override
    public List<Location> discover(List<String> attachedLocations, Map<String, String> args) {
        return List.of(
            new Location("test_file_1.parquet", true, "parquet", "v1"),
            new Location("test_file_2.parquet", true, "parquet", "v1")
        );
    }

    @Override
    public VectorSchemaRoot data(Location location, Map<String, String> args) {
        Schema schema = new Schema(Arrays.asList(
            Field.nullable("id", new ArrowType.Int(64, true)),
            Field.nullable("name", new ArrowType.Utf8())
        ));

        VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);

        switch (location.location()) {
            case "test_file_1.parquet" -> {
                root.allocateNew();
                BigIntVector idVec = (BigIntVector) root.getVector("id");
                VarCharVector nameVec = (VarCharVector) root.getVector("name");
                idVec.setSafe(0, 1); nameVec.setSafe(0, "alice".getBytes());
                idVec.setSafe(1, 2); nameVec.setSafe(1, "bob".getBytes());
                idVec.setSafe(2, 3); nameVec.setSafe(2, "charlie".getBytes());
                root.setRowCount(3);
            }
            case "test_file_2.parquet" -> {
                root.allocateNew();
                BigIntVector idVec = (BigIntVector) root.getVector("id");
                VarCharVector nameVec = (VarCharVector) root.getVector("name");
                idVec.setSafe(0, 4); nameVec.setSafe(0, "dave".getBytes());
                idVec.setSafe(1, 5); nameVec.setSafe(1, "eve".getBytes());
                root.setRowCount(2);
            }
            default -> {
                root.close();
                return null;
            }
        }
        return root;
    }

    public static void main(String[] args) {
        Serve.run(new TestSource());
    }
}

File: sdk/rust/examples/test_source.rs
use arrow::array::{Int64Array, StringArray};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use bundlebase_sdk::{Connector, Location};
use std::collections::HashMap;
use std::sync::Arc;

struct TestSource;

impl Connector for TestSource {
    fn discover(
        &self,
        _attached: &[String],
        _args: &HashMap<String, String>,
    ) -> Result<Vec<Location>, Box<dyn std::error::Error>> {
        Ok(vec![
            Location {
                location: "test_file_1.parquet".into(),
                must_copy: true,
                format: "parquet".into(),
                version: "v1".into(),
                num_rows: Some(3),
            },
            Location {
                location: "test_file_2.parquet".into(),
                must_copy: true,
                format: "parquet".into(),
                version: "v1".into(),
                num_rows: Some(2),
            },
        ])
    }

    fn data(
        &self,
        location: &Location,
        _args: &HashMap<String, String>,
    ) -> Result<Option<Vec<RecordBatch>>, Box<dyn std::error::Error>> {
        let schema = Arc::new(Schema::new(vec![
            Field::new("id", DataType::Int64, false),
            Field::new("name", DataType::Utf8, false),
        ]));

        match location.location.as_str() {
            "test_file_1.parquet" => {
                let batch1 = RecordBatch::try_new(
                    schema.clone(),
                    vec![
                        Arc::new(Int64Array::from(vec![1, 2])),
                        Arc::new(StringArray::from(vec!["alice", "bob"])),
                    ],
                )?;
                let batch2 = RecordBatch::try_new(
                    schema,
                    vec![
                        Arc::new(Int64Array::from(vec![3])),
                        Arc::new(StringArray::from(vec!["charlie"])),
                    ],
                )?;
                Ok(Some(vec![batch1, batch2]))
            }
            "test_file_2.parquet" => {
                let batch = RecordBatch::try_new(
                    schema,
                    vec![
                        Arc::new(Int64Array::from(vec![4, 5])),
                        Arc::new(StringArray::from(vec!["dave", "eve"])),
                    ],
                )?;
                Ok(Some(vec![batch]))
            }
            _ => Ok(None),
        }
    }
}

fn main() {
    bundlebase_sdk::serve(&TestSource);
}
